sklearn/isotonic.py

```diff
@@ -121,10 +122,8 @@ def isotonic_regression(y, sample_weight=None, y_min=None, y_max=None,
     order = np.s_[:] if increasing else np.s_[::-1]
     y = check_array(y, ensure_2d=False, dtype=[np.float64, np.float32])
     y = np.array(y[order], dtype=y.dtype)
-    if sample_weight is None:
-        sample_weight = np.ones(len(y), dtype=y.dtype)
-    else:
-        sample_weight = np.array(sample_weight[order], dtype=y.dtype)
+    sample_weight = _check_sample_weight(sample_weight, y, dtype=y.dtype)
+    sample_weight = np.ascontiguousarray(sample_weight[order])
```

Comment:
```suggestion
    sample_weight = sample_weight[order]
```

That's a 1D array so it has to be contiguous

Reply:
is it worth adding a parameter `order` to `_check_sample_weight` in order to avoid a copy when sample_weight is None ?

Reply:
It seem that reordering a 1D array with a reverse index expression does not produce a contiguous array:
```
_______________________________________ test_isotonic_ymin_ymax _______________________________________ 
                                                                                                        
    def test_isotonic_ymin_ymax():                                                                      
        # Test from @NelleV's issue:                                                                    
        # https://github.com/scikit-learn/scikit-learn/issues/6921                                      
        x = np.array([1.263, 1.318, -0.572, 0.307, -0.707, -0.176, -1.599, 1.059,                       
                      1.396, 1.906, 0.210, 0.028, -0.081, 0.444, 0.018, -0.377,                         
                      -0.896, -0.377, -1.327, 0.180])                                                   
        y = isotonic_regression(x, y_min=0., y_max=0.1)                                                 
                                                                                                        
        assert np.all(y >= 0)                                                                           
        assert np.all(y <= 0.1)                                                                         
                                                                                                        
        # Also test decreasing case since the logic there is different                                  
>       y = isotonic_regression(x, y_min=0., y_max=0.1, increasing=False)                               
                                                                                                        
sklearn\tests\test_isotonic.py:394:                                                                     
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sklearn\isotonic.py:128: in isotonic_regression                                                         
    _inplace_contiguous_isotonic_regression(y, sample_weight)                                           
sklearn\_isotonic.pyx:15: in sklearn._isotonic._inplace_contiguous_isotonic_regression                  
    def _inplace_contiguous_isotonic_regression(floating[::1] y, floating[::1] w):                      
stringsource:658: in View.MemoryView.memoryview_cwrapper                                                
    ???                                                                                                 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
                                                                                                        
>   ???                                                                                                 
E   ValueError: ndarray is not C-contiguous                                                             
```

With my debugger on, I have the confirmation that with `increasing=False` the `sample_weight[order].flags` has `c_contiguous=False`

Reply:
Right, `a[::-1]` is a view and thus is not contiguous

Reply:
> is it worth adding a parameter order to _check_sample_weight in order to avoid a copy when sample_weight is None ?

I'm not sure what you mean, but related, I have proposed to add a `return_ones` parameter in the past to avoid returning an array of ones when sample_weight is None.

Reply:
> That's a 1D array so it has to be contiguous

Not necessarily:

```
>>> a = np.ones(100)
>>> b = a[::2]
>>> b.flags
  C_CONTIGUOUS : False
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False

```

Reply:
what would return_ones return ?

Reply:
Fair, but in this specific case the slices are `::1` or `::-1`, so the view is contiguous even though numpy does't see it as such

Reply:
Indeed any kind of strides will make the array non-contiguous.

Reply:
to summarize:
- view of 1D array will likely not be contiguous.
- fancy indexing always copy so the new array is contiguous.

here it can be a view so we need to make it contiguous.

Reply:
>  what would return_ones return ?

np.ones if True, else None.

This avoids protecting the call with

```py
if sample_weight is not None:
	sample_weight = _check_sample_weight...
```

which is needed in some algorithms that support SW but have a fast path when sample_weight is None

Reply:
> Fair, but in this specific case the slices are ::1 or ::-1, so the view is contiguous even though numpy does't see it as such

for ::-1 no because contiguous means stride of +1 not stride of + or - 1

Reply:
> which is needed in some algorithms that support SW but have a fast path when sample_weight is None

good point, can you give me the link to the corresponding discussion ?

Reply:
> for ::-1 no because contiguous means stride of +1 not stride of + or - 1

Do you have a ref regarding the stride?

> good point, can you give me the link to the corresponding discussion ?

https://github.com/scikit-learn/scikit-learn/issues/15358#issuecomment-549051299

Reply:
> Do you have a ref regarding the stride?

Hum no :/ but my understanding was that the order should determine uniquely how to loop through the array.

In the case of this PR, it's actually enforced in cython: `floating[::1] w`

Reply:
ok yeah makes sense